.github/workflows/code.yml

```
@@ -17,24 +17,34 @@ jobs:
       - name: Checkout
         uses: actions/checkout@v3
 
-      - name: Setup python
-        uses: actions/setup-python@v4
+      - name: Install
+        uses: ./.github/actions/install_requirements
         with:
-          python-version: "3.10"
+          requirements_file: requirements-lint.txt
+          install_options: -e .[dev]
 
       - name: Lint
-        run: |
-          touch requirements_dev.txt requirements.txt
-          pip install -r requirements.txt -r requirements_dev.txt -e .[dev]
-          tox -e pre-commit,mypy
+        run: tox -e pre-commit,mypy
+
+      - name: Upload lockfiles
+        uses: actions/upload-artifact@v3
+        with:
+          name: lockfiles
+          path: lockfiles
 
   test:
     if: github.event_name != 'pull_request' || github.event.pull_request.head.repo.full_name != github.repository
     strategy:
       fail-fast: false
       matrix:
         os: ["ubuntu-latest"] # can add windows-latest, macos-latest
-        python: ["3.8", "3.9", "3.10"]
+        python: ["3.8", "3.9", "3.10", "3.11"]
```

Oops, that makes a duplicate job
```suggestion
        python: ["3.9", "3.10", "3.11"]
```